### PR TITLE
Implement hash join executor support

### PR DESCRIPTION
**Pull request by panarch**
Implement hash join executor support which can boost execution speed of some join queries with `key = value` shape constraints.
Hash join executor loads all data to the memory by building hash map based on evaluating pre-planned key expr.
It can dramatically reduce the overhead of accessing storage and join constraint evaluation time.

* Add JoinExecutor field to ast::Join.
`key_expr`: It used as key of HashMap, evaluation is done when executor builds the map.
`value_expr`: value of HashMap, evaluated in the loop of join execution.
`where_clause`: filtering expr which can reduce the size of the HashMap.

* Add join planner module.
`plan/mock.rs` - MockStorage impl which is only for testing planners.
`plan/schema.rs` - Analyze AST and returns all related table schema list.
`plan/evaluable.rs` - Check whether the AST (query) is evaluable of not.
`plan/join.rs` - Join Planner, currently the only role is to find HASH JOIN.

### CODE REVIEW

**Review by ever0de — COMMENTED** [review]
I leave a review (nit-picking) that has been merged.

File: core/src/plan/evaluable.rs [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+        }
+        PlanExpr::Expr(expr) => check_expr(context, expr),
+        PlanExpr::TwoExprs(expr, expr2) => {
+            check_expr(context.as_ref().map(Rc::clone), expr) && check_expr(context, expr2)
+        }
+        PlanExpr::ThreeExprs(expr, expr2, expr3) => {
+            check_expr(context.as_ref().map(Rc::clone), expr)
+                && check_expr(context.as_ref().map(Rc::clone), expr2)
+                && check_expr(context, expr3)
+        }
```
[the remark this answers is not in the document]
Comment (panarch) [reply core/src/plan/evaluable.rs:37]: `MultiExprs` always uses `Vec`.
This enums are for avoiding unnecessary heap uses, to use less boxes.

File: core/src/executor/join/mod.rs [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+        let init_rows: Joined<'a> = Box::pin(rows.map(|row| row.map(Rc::new)));
+        let joins = self
+            .join_clauses
+            .iter()
+            .enumerate()
+            .map(move |(i, join_clause)| {
+                let join_columns = Rc::clone(&self.join_columns[i]);
+
+                Ok::<_, Error>((join_clause, join_columns))
+            });
```
[the remark this answers is not in the document]
Comment (panarch) [reply core/src/executor/join/mod.rs:67]: Oh, I wasn't aware of this. `zip` looks much better.
Thanks a lot. I already merged this so I'll make a new branch to apply this, thanks!

File: core/src/executor/join/mod.rs [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+            join_columns,
+            filter_context,
+        }
+    }
+
+    pub async fn apply(
+        self,
+        rows: impl Stream<Item = Result<BlendContext<'a>>> + 'a,
+    ) -> Result<Joined<'a>> {
+        let init_rows: Joined<'a> = Box::pin(rows.map(|row| row.map(Rc::new)));
```
[the remark this answers is not in the document]
Comment (panarch) [reply core/src/executor/join/mod.rs:58]: I didn't know that lifetime in here can be omitted, thanks!

File: core/src/executor/join/hash_key.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+pub enum HashKey {
+    I8(i8),
+    I64(i64),
+    Bool(bool),
+    Str(String),
+    Date(NaiveDate),
+    Timestamp(NaiveDateTime),
+    Time(NaiveTime),
+    Interval(Interval),
+    Uuid(u128),
```
[the remark this answers is not in the document]
Comment (panarch) [reply core/src/executor/join/hash_key.rs:22]: This is simple. `Value` enum also contains `Uuid` with `u128` type.
`UUID` can be described as `[u8; 16]` which is identical to `u128`.

File: core/src/executor/join/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (panarch) [reply core/src/executor/join/mod.rs:67]: Thanks, I cleaned a bit more on it.
```rust
..
let joins = self                                              
    .join_clauses                                             
    .iter()                                                   
    .zip(self.join_columns.iter().map(Rc::clone));            
                                                              
stream::iter(joins)                                           
    .map(Ok)                                                  
    .try_fold(init_rows, |rows, (join_clause, join_columns)| {
..
```

**Pull request merged**